app/code/Magento/Sales/Model/InvoiceOrder.php

```diff
@@ -139,7 +139,8 @@ public function execute(
         InvoiceCreationArgumentsInterface $arguments = null
     ) {
         $connection = $this->resourceConnection->getConnection('sales');
-        $order = $this->orderRepository->get($orderId);
+        $connection->beginTransaction();
+        $order = $this->orderRepository->getForUpdate($orderId);
```

Comment:
Unfortunately, we can't accept such contribution to the core, because it violates several OOP principles, as well as it contradicts the way how we  (Magento an Adobe) see the future evolution of the system to be distributed deployable in the cloud. 

First, let me make a touch base regarding the OOP principle violations:
1. We can't introduce method `getForUpdate` to the Repository if the repository does not guarantee locking of the entity. "Select for Update" would not make a lot of sense if it's called outside the transition. That's why you should have:
```php
	        $connection->beginTransaction();
            $order = $this->orderRepository->getForUpdate($orderId);
```
to make it work.
This is called - Leaky Abstraction https://www.joelonsoftware.com/2002/11/11/the-law-of-leaky-abstractions/
the low-level detail of how to retrieve the data from DB leaked to high-level interface of Repository, which supposed to provide persistence agnostic behavior for business logic working with an entire. 
Another thing is that you have a tandem of 2 different methods need to be called together, and it would be considered a logical error, if `$this->orderRepository->getForUpdate($orderId);` will be called outside of the transaction scope. This says about wrong object decomposition, as high-level service Repository is dependent on low-level Connection object beginTransaction to be called beforehand. 

Besides OOP principles and talking about distributed systems and requirements for them, we could say that strict consistency is very costly to support, and its cost grows proportionally with the size of an application. Thus, transactional logic and "Select .. for update" might work perfectly well for the small application deployed on the single box server, while as soon as we are dealing with microservices deployed separately (each relying on own data storage) - it's incredibly hard to guarantee strict consistency. 2-phase/3-phase commit protocols are proven their misconception in modern applications. Thus, the only viable option is to use Optimistic locking strategy (rather than pessimistic one) with an ability to rollback changes applied via mechanism of compensation transactions or Sagas https://blog.couchbase.com/saga-pattern-implement-business-transactions-using-microservices-part/

Reply:
Potential solution to the leaky abstraction is to encapsulate the transaction in the repository, for example:
```php
public function getForUpdate(int $orderId, callable $callback)
{
    $this->startTransaction();
    try {
        $order = $this->loadForUpdate($orderId);
        $callback($order);
        $this->commit();
    } catch (\Throwable $e) {
        $this->rollback();
        throw $e;
    }
}
```
```php
$this->orderRepository->getForUpdate($orderId, function ($order) {
    // Update order within transaction...
    $this->orderRepository->save($order);
});
```

Admittedly, the JavaScript-like callback does not fit well within the Magento framework.

Reply:
@maghamed fair point, @SerhiyShkolyarenko @0x15f  can we consider something like this, https://help.sap.com/doc/a4265d5ea8314eb2929e6cf6fb8e35a5/1811/en-US/de/hybris/platform/tx/Transaction.html#lock-de.hybris.platform.jalo.Item- since the services aware about transactions it should be OK

Reply:
@maghamed If I refractor the PR to match the implementation @akaplya referenced above and then implement the sales order locking as a configurable option, would it then be accepted?

Reply:
@maghamed, @kandy , it may work, since the transactions already present on this tier, no changes on contract level require. Should be added a new method into API class, which does not suppose to have an override, or even a new class introduced. The behavior could be optional, and enabled only by the merchant decision, disabled by default. This is not as good as optimistic locking, but easier to deliver.

`
	$connection->beginTransaction();
	$order = $this->orderRepository->get($orderId);
	$connection->lock($order);
	// OR $lockManager->lock($order); -- btw lock manager could assume different implementations.
        try {
	    // do something
            $connection->commit();
        } catch (\Exception $e) {
            $connection->rollBack();
        }

Reply:
@akaplya I like that approach, to increase the performance of it we can select only the primary key and return true/false based on the error state. I think a LockManagerInterface is the way to go, this would allow other implementations with a core locking principle implemented.

Reply:
I like the solution proposed by @akaplya.

Reply:
The @akaplya's pseudo-code above does look good. However, and maybe I'm missing something, how can such a code possibly work given it separates the read operation from the lock? Are you suggesting that all other processes unable to obtain the lock error out after successfully reading an outdated version of the data being updates by the process holding the lock? If so, this approach is significantly inferior to the behavior of `UPDATE..FOR SELECT` (that combines read and lock into a single atomic operation) that makes the concurrent processes wait for the updated data to become available w/o throwing errors.

Reply:
@sshymko As for me, the main issue that we already implicitly have a lock, but it happens after the business-critical operations were performed, status calculated, shipment integration invoked, payment processed, so on. This code moves the lock on the level before any of the state modifying operations happened. 

but, I see your point - due to order has a state, the state after lock could be different, this is a valid concern. In the perfect world, it should not affect the status calculation and status should not be embedded in order. but in the real world, we can slightly rewrite the approach.

```
	$connection->beginTransaction();
	$lockManager->lock($order::class, $orderId);
	$order = $this->orderRepository->get($orderId);
	try {
	    // do something
            $connection->commit();
        } catch (\Exception $e) {
            $connection->rollBack();
        }
```

Reply:
@akaplya 
This should work although it's not as pleasing to work with as the previous version. A developer would now be able to accidentally make a mistake/typo and mismatch argument of `lock()` and `get()`.

Also, how are you proposing to release the lock? Implicitly on `commit()` and `rollback()`? As @maghamed originally pointed out, that would create a leaky abstraction between the lock mechanism and the transactions. So, an explicit lock release would be probably necessary. Once again, such design puts responsibility of maintaining symmetry of the `lock/unlock()` calls on a developer opening the possibility for mistakes.

Proposed implementation:
```php
$connection->beginTransaction();
$lock = $lockManager->lock(Order::class, $orderId);
$order = $this->orderRepository->get($orderId);
try {
    // do something
    $connection->commit();
} catch (\Throwable $e) {
    $connection->rollBack();
} finally {
    $lockManager->release($lock);
}
```

Reply:
That good that we dive into implementation details. However, in my opinion, the lock should happen on the lower layer, similar to the connection. It should not leak to the business layer, which, in our case, is represented with the repository. Implicit abstraction may play a harmful role here. The implementation which is responsible for setting lock should be explicitly related to the connection to avoid any misguidance of how this lock will be released.
Probably the next semantic could be more clear.
`$lockManager->lock($connection, Order::class, $orderId);`
My point here, to make it as much explicit as possible to the developer that the lock belongs to the transaction until we do not have BPMN or similar to implement compensation action.

Reply:
I see `\Magento\Framework\Lock\Backend\Database` already has a locking mechanism using `GET_LOCK()`. Should this be implemented as a Resource/Object Lock Manager to avoid confusion with a duplicate lock manager? I believe taking that approach may also eliminate confusion when overriding the lock manager.